File: public/ha-addon/stockjp-print/agent.mjs
#!/usr/bin/env node
/**
 * Agent d'impression local — Stock JP/JC → Brother QL-810Wc
 *
 * Scrute la file print_jobs côté cloud, télécharge le PDF et l'envoie à
 * l'imprimante locale via CUPS (lp) sur macOS/Linux ou PowerShell sur Windows.
 *
 * Usage :
 *   1. cp config.example.json config.json
 *   2. Remplir : SUPABASE_URL, SUPABASE_ANON_KEY, EMAIL, PASSWORD, PRINTER
 *   3. node agent.mjs
 */
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { spawn } from "node:child_process";
import { createClient } from "@supabase/supabase-js";

const cfgPath = path.resolve(process.cwd(), "config.json");
if (!fs.existsSync(cfgPath)) {
  console.error("config.json introuvable. Copiez config.example.json → config.json et remplissez-le.");
  process.exit(1);
}
const cfg = JSON.parse(fs.readFileSync(cfgPath, "utf-8"));
const supabase = createClient(cfg.SUPABASE_URL, cfg.SUPABASE_ANON_KEY, {
  auth: { persistSession: false },
});

console.log("→ Connexion…");
const { error: authErr } = await supabase.auth.signInWithPassword({
  email: cfg.EMAIL, password: cfg.PASSWORD,
});
if (authErr) { console.error("Auth :", authErr.message); process.exit(1); }
console.log(`✓ Connecté · imprimante cible: ${cfg.PRINTER}`);

const POLL_MS = cfg.POLL_MS ?? 4000;
const isWin = os.platform() === "win32";

// Mapping format → option media CUPS (doit correspondre à src/lib/print.ts ROLL_SPECS)
const CUPS_MEDIA = {
  "23x23":  "Custom.23x23mm",
  "17x54":  "Custom.17x54mm",
  "62x29":  "Custom.29x62mm",
  "62x100": "Custom.62x100mm",
  "62":     "Custom.62x30mm",
  // tolérance ancien libellé
  "30x62":  "Custom.62x30mm",
  "62x30":  "Custom.62x30mm",
};

async function printPdf(pdfBuffer, jobId, format) {
  const tmp = path.join(os.tmpdir(), `stockjp-${jobId}.pdf`);
  fs.writeFileSync(tmp, pdfBuffer);
  const media = CUPS_MEDIA[format] ?? "Custom.62x30mm";
  return new Promise((resolve, reject) => {
    let cmd, args;
    if (isWin) {
      cmd = "powershell";
      args = ["-Command", `Start-Process -FilePath '${tmp}' -Verb Print -PassThru | %{ Start-Sleep 4; $_ } | kill`];
    } else {
      cmd = "lp";
      args = ["-d", cfg.PRINTER, "-o", `media=${media}`, "-o", "fit-to-page", tmp];
    }
    const p = spawn(cmd, args, { stdio: "inherit" });
    p.on("close", (code) => {
      try { fs.unlinkSync(tmp); } catch {}
      code === 0 ? resolve() : reject(new Error(`${cmd} a retourné ${code}`));
    });
  });
}

async function pollOnce() {
  const { data: jobs, error } = await supabase
    .from("print_jobs")
    .select("*")
    .eq("status", "pending")
    .order("created_at", { ascending: true })
    .limit(5);
  if (error) { console.error("Poll :", error.message); return; }
  for (const j of jobs ?? []) {
    console.log(`→ Job ${j.id.slice(0, 8)} (${j.format})`);
    await supabase.from("print_jobs").update({ status: "printing" }).eq("id", j.id);
    try {
      if (!j.pdf_base64) throw new Error("PDF absent");
      const buf = Buffer.from(j.pdf_base64, "base64");
      await printPdf(buf, j.id, j.format);
      await supabase.from("print_jobs").update({
        status: "printed", printed_at: new Date().toISOString(), error: null,
      }).eq("id", j.id);
      console.log(`  ✓ imprimé`);
    } catch (e) {
      console.error(`  ✗ ${e.message}`);
      await supabase.from("print_jobs").update({
        status: "error", error: String(e.message ?? e),
      }).eq("id", j.id);
    }
  }
}

console.log(`Polling toutes les ${POLL_MS}ms…`);
setInterval(pollOnce, POLL_MS);
pollOnce();

File: public/print-agent-py/agent.py
#!/usr/bin/env python3
"""
Agent d'impression Stock JP/JC -> Brother QL-810W (IP directe, port 9100)

Pipeline :
  1. Poll Supabase pour les print_jobs en status 'pending'
  2. Décode le PDF (base64) et le rastérise en image 1-bit via PyMuPDF
  3. Convertit en commandes raster Brother via brother_ql
  4. Envoie le flux binaire en TCP direct à l'imprimante (tcp://IP:9100)

Aucune installation de driver côté Windows requise.

Usage :
  pip install -r requirements.txt
  cp config.example.json config.json
  # éditer config.json (SUPABASE_URL/KEY, EMAIL/PASSWORD, PRINTER_IP)
  python agent.py
"""
import base64
import io
import json
import os
import sys
import time
from pathlib import Path

import fitz  # PyMuPDF
from PIL import Image
from supabase import create_client
from brother_ql.raster import BrotherQLRaster
from brother_ql.conversion import convert
from brother_ql.backends.helpers import send

CFG_PATH = Path(__file__).parent / "config.json"
if not CFG_PATH.exists():
    print("config.json introuvable. Copiez config.example.json -> config.json")
    sys.exit(1)

cfg = json.loads(CFG_PATH.read_text(encoding="utf-8"))

SUPABASE_URL = cfg["SUPABASE_URL"]
SUPABASE_KEY = cfg["SUPABASE_ANON_KEY"]
EMAIL = cfg["EMAIL"]
PASSWORD = cfg["PASSWORD"]
PRINTER_IP = cfg["PRINTER_IP"]
MODEL = cfg.get("MODEL", "QL-810W")
POLL_MS = cfg.get("POLL_MS", 4000)
DPI = cfg.get("DPI", 300)  # 300 = standard Brother QL

# Mapping format app -> label brother_ql
# voir https://github.com/pklaus/brother_ql#supported-media
LABELS = {
    "23x23":  "23x23",
    "17x54":  "17x54",
    "62x29":  "29x62",
    "62x100": "62x100",
    "62":     "62",        # rouleau continu 62 mm
    "29x50":  "29",        # rouleau continu 29 mm (DK-22211)
    "50x29":  "29",
    "30x62":  "62",
    "62x30":  "62",
}

sb = create_client(SUPABASE_URL, SUPABASE_KEY)
print("-> Connexion...")
auth = sb.auth.sign_in_with_password({"email": EMAIL, "password": PASSWORD})
if not auth.user:
    print("Auth échouée")
    sys.exit(1)
print(f"OK  Connecté · imprimante tcp://{PRINTER_IP}:9100 · modèle {MODEL}")


def pdf_to_images(pdf_bytes: bytes) -> list[Image.Image]:
    """Rastérise chaque page du PDF en image PIL (mode RGB)."""
    doc = fitz.open(stream=pdf_bytes, filetype="pdf")
    images = []
    for page in doc:
        # zoom = DPI / 72 (PDF est en 72 dpi)
        mat = fitz.Matrix(DPI / 72, DPI / 72)
        pix = page.get_pixmap(matrix=mat, alpha=False)
        img = Image.frombytes("RGB", (pix.width, pix.height), pix.samples)
        images.append(img)
    doc.close()
    return images


def print_pdf(pdf_b64: str, fmt: str) -> None:
    label = LABELS.get(fmt)
    if not label:
        raise RuntimeError(f"Format d'étiquette non supporté par l'agent: {fmt}")
    pdf_bytes = base64.b64decode(pdf_b64)
    pages = pdf_to_images(pdf_bytes)
    if not pages:
        raise RuntimeError("PDF sans page")

    qlr = BrotherQLRaster(MODEL)
    qlr.exception_on_warning = True

    # brother_ql redimensionne et tourne automatiquement selon le label
    instructions = convert(
        qlr=qlr,
        images=pages,
        label=label,
        rotate="auto",
        threshold=70.0,
        dither=False,
        compress=False,
        red=False,
        dpi_600=False,
        hq=True,
        cut=True,
    )
    send(
        instructions=instructions,
        printer_identifier=f"tcp://{PRINTER_IP}:9100",
        backend_identifier="network",
        blocking=True,
    )


def poll_once() -> None:
    try:
        res = (
            sb.table("print_jobs")
            .select("*")
            .eq("status", "pending")
            .order("created_at", desc=False)
            .limit(5)
            .execute()
        )
    except Exception as e:
        print(f"Poll: {e}")
        return

    for j in res.data or []:
        jid = j["id"]
        fmt = j["format"]
        print(f"-> Job {jid[:8]} ({fmt})")
        sb.table("print_jobs").update({"status": "printing"}).eq("id", jid).execute()
        try:
            if not j.get("pdf_base64"):
                raise RuntimeError("PDF absent")
            print_pdf(j["pdf_base64"], fmt)
            sb.table("print_jobs").update({
                "status": "printed",
                "printed_at": "now()",
                "error": None,
            }).eq("id", jid).execute()
            print("   OK imprimé")
        except Exception as e:
            msg = str(e)
            print(f"   FAIL {msg}")
            sb.table("print_jobs").update({
                "status": "error",
                "error": msg[:500],
            }).eq("id", jid).execute()


print(f"Polling toutes les {POLL_MS}ms... (Ctrl+C pour arrêter)")
try:
    while True:
        poll_once()
        time.sleep(POLL_MS / 1000)
except KeyboardInterrupt:
    print("\nArrêt.")
